_posts/2017-11-30-totw-119.md

```diff
@@ -0,0 +1,264 @@
+    qualified names (with leading `::`) unless you are referring to a name
+    inside the current namespace.
+
+*   For other uses of names, avoid fully qualifying when reasonable; see
+    [TotW 130](/tips/130).
+
+(Remember that you can always have a local namespace alias or
+*using-declaration* in a block scope, which can be handy in header-only
+libraries.)
+
+## Background
+
+C++ organizes names into *namespaces*. This crucial facility allows code bases
+to scale by keeping ownership of names local and not requiring non-local
+agreement on name use. However, namespaces impose a certain cosmetic burden,
+since qualified names (`foo::Bar`) are often long and quickly become clutter. We
+often find it convenient to use *unqualified names* (`Bar`). Additionally, we
+may wish to introduce a namespace alias for a long but frequently used
+namespace: `namespace eu = example::v1::util;` We will collectively call
+<em>using-declaration</em>s and namespace aliases just *aliases* in this tip.
+
+## The problem
+
+The purpose of a namespace is to help code authors avoid name collisions, both
+at the point of name lookup and at the point of linking. Aliases can potentially
+undermine the protection afforded by namespaces. The problem has two separate
+aspects: the scope of the alias, and the use of relative qualifiers.
+
+### Scope of the alias
+
+The scope at which you place an alias can have subtle effects on code
+maintainability. Consider the following two variants:
+
+```c++
+using ::foo::Quz;
+
+namespace example {
+namespace util {
+
+using ::foo::Bar;
+```
+
+It appears that both <em>using-declaration</em>s are effective at making the
+names `Bar` and `Quz` available for unqualified lookup inside our working
+namespace `::example::util`. For `Bar`, everything is working as expected,
+provided there is no other declaration of `Bar` inside namespace
+`::example::util`. But this is your namespace, so it is in your power to control
+this.
+
+On the other hand, if a header is later included that declares a global name
+`Quz`, then the first *using-declaration* becomes ill-formed, as it attempts to
+redeclare the name `Quz`. And if another header declares `::example::Quz` or
+`::example::util::Quz`, then unqualified lookup will find _that_ name rather
+than your alias.
+
+This brittleness can be avoided if you do not add names to namespaces that you
+do not own (which includes the global namespace). By placing the aliases inside
+your own namespace, the unqualified lookup finds your alias first and never
+continues to search containing namespaces.
+
+More generally, the closer a declaration is to the point of use, the smaller is
+the set of scopes that can break your code. At the worst end of our example is
+`Quz`, which can be broken by anyone; `Bar` can only be broken by other code in
+`::example::util`, and a name that is declared and used inside an unnamed
+namespace cannot be broken by any other scope ([see below](#unnamed-namespaces)
```

Comment:
[see below] ... -> See [Unnamed Namespaces](#unnamed-namespaces) below for an example.

(Always name a link, usually the heading or title, and initial caps it as a heading.)